libs/langchain/langchain/retrievers/google_vertex_ai_search.py

```diff
@@ -419,17 +429,15 @@ def _get_relevant_documents(
         )
 
         request = ConverseConversationRequest(
-            name=self._client.conversation_path(
-                self.project_id,
-                self.location_id,
-                self.data_store_id,
-                self.conversation_id,
-            ),
+            name=self.conversation_name,
             serving_config=self._serving_config,
             query=TextInput(input=query),
         )
         response = self._client.converse_conversation(request)
 
+        # Preserve Conversation State between requests
+        self.conversation_name = response.conversation.name
```

Comment:
will response.conversation_name ever be different than request.name?

Reply:
It will be if `request.name` has conversation_id `-`. (Which means: "start a new conversation").

Then `response.conversation.name` will be a randomly generated conversation id that can be sent in subsequent requests to continue that same conversation session.

The official code sample is in this documentation

https://cloud.google.com/generative-ai-app-builder/docs/multi-turn-search

@baskaryan

Reply:
I'm actually not sure whether the client should be stateful. This is a retriever and state preservation will make this confusing. I'd suggest to provide an optional `conversation_name` as an arg to the `_get_relevant_documents`.

Otherwise using this retriever might lead to mistakes.

Reply:
Agreed with @lkuligin . Especially in environments like LangServe, a user might expect the behavior to match a jupyter notebook, and this would cause a potentially unexpected error, where in a notebook it would retain the `conversation_name` state, but in LangServe it wouldn't for a second turn of conversation.

Passing it in would force folks to record that state on their own and handle accordingly.

Reply:
I understand @lkuligin's suggestion about not making the client stateful in itself.

However, with the layers of abstraction required for a retriever, I'm not sure of the best way to output the conversation_id generated in the first request so that the user can persist it. Since `_get_relevent_documents()` only outputs documents with their own individual metadata.

Reply:
Going to propose something different in a new PR shortly. The issue is that Retrievers need to be stateless, and this change introduces state into it.

The better alternative is to get this ID from a config, like the `RunnableWithmessageHistory`: https://python.langchain.com/docs/expression_language/how_to/message_history#invoking-with-config